# Hypothesis Testing

The purpose of the test is to tell if there is any significant difference between two data sets.



## Overview

This module covers,

1) One sample and Two sample t-tests

2) ANOVA

3) Type I and Type II errors

4) Chi-Squared Tests

## Question 1 

*A student is trying to decide between two GPUs. He want to use the GPU for his research to run Deep learning algorithms, so the only thing he is concerned with is speed.*

*He picks a Deep Learning algorithm on a large data set and runs it on both GPUs 15 times, timing each run in hours. Results are given in the below lists GPU1 and GPU2.*

In [64]:
from scipy import stats 
import numpy as np

In [65]:
GPU1 = np.array([11,9,10,11,10,12,9,11,12,9,11,12,9,10,9])
GPU2 = np.array([11,13,10,13,12,9,11,12,12,11,12,12,10,11,13])

#Assumption: Both the datasets (GPU1 & GPU 2) are random, independent, parametric & normally distributed

Hint: You can import ttest function from scipy to perform t tests 

**First T test**

*One sample t-test*

Check if the mean of the GPU1 is equal to zero.
- Null Hypothesis is that mean is equal to zero.
- Alternate hypothesis is that it is not equal to zero.

In [66]:
print ("Length of GPU1 Samples : " , len(GPU1) , " GPU1 Mean : ", GPU1.mean())
print ("Length of GPU2 Samples : " , len(GPU2) , " GPU2 Mean : ", GPU2.mean())
print ("Degress of Freedom     : ", len(GPU1)+len(GPU2)-1)

Length of GPU1 Samples :  15  GPU1 Mean :  10.333333333333334
Length of GPU2 Samples :  15  GPU2 Mean :  11.466666666666667
Degress of Freedom     :  29


Mean of GPU1 is not 0

### 1-Sample T-Test

In [67]:
# 1 Sample T-Test for GPU1 with population mean == 0 
tt_1Sample = stats.ttest_1samp(GPU1, 0)
tt_1Sample

Ttest_1sampResult(statistic=34.056241516158195, pvalue=7.228892044970457e-15)

In [68]:
# 1 Sample T-Test for GPU2 with population mean == 0

tt_1Sample2 = stats.ttest_1samp(GPU2, 0)
tt_1Sample2

Ttest_1sampResult(statistic=37.406459913272236, pvalue=1.969680571681882e-15)

## Question 2

Given,

Null Hypothesis : There is no significant difference between data sets

Alternate Hypothesis : There is a significant difference

*Do two-sample testing and check whether to reject Null Hypothesis or not.*

GPU Tests 

* $H_0$: Null Hypothesis      : "No Significate diff between sample data sets" .
  * $\mu_1$ - $\mu_2$ = 0

* $H_A$: Alternate Hypothesis : Rejected Null Hypothesis. Significate diff between sample data sets

https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html

In [69]:
print("Diff of means of GPU1 and GPU2 : ", GPU1.mean() - GPU2.mean())

Diff of means of GPU1 and GPU2 :  -1.1333333333333329


In [70]:
# tt_gpu = stats.ttest_ind(GPU1, GPU2)
tt_ind = ttest_ind(GPU1, GPU2)
tt_ind

Ttest_indResult(statistic=-2.627629513471839, pvalue=0.013794282041452725)

In [71]:
# Degree of Freedom 

deg_freedom_gpu = len(GPU1) - 1 + len(GPU2) - 1
deg_freedom_gpu


28

t-Value = -2.62 with pValue = 1.39 (Less than 2.5) Hence, we faile to reject the Null Hypothesis.  

Both GPU1 and GPU2 have comparable mean. 

## Question 3

He is trying a third GPU - GPU3.

In [72]:
GPU3 = np.array([9,10,9,11,10,13,12,9,12,12,13,12,13,10,11])

#Assumption: Both the datasets (GPU1 & GPU 3) are random, independent, parametric & normally distributed

*Do two-sample testing and check whether there is significant differene between speeds of two GPUs GPU1 and GPU3.*

#### Answer:

In [73]:
# tt_gpu = stats.ttest_ind(GPU1, GPU2)
tt_ind3 = ttest_ind(GPU1, GPU3)
tt_ind3

Ttest_indResult(statistic=-1.4988943759093303, pvalue=0.14509210993138993)

## ANOVA

## Question 4 

If you need to compare more than two data sets at a time, an ANOVA is your best bet. 

*The results from three experiments with overlapping 95% confidence intervals are given below, and we want to confirm that the results for all three experiments are not significantly different.*

But before conducting ANOVA, test equality of variances (using Levene's test) is satisfied or not. If not, then mention that we cannot depend on the result of ANOVA

In [74]:
import numpy as np

e1 = np.array([1.595440,1.419730,0.000000,0.000000])
e2 = np.array([1.433800,2.079700,0.892139,2.384740])
e3 = np.array([0.036930,0.938018,0.995956,1.006970])

#Assumption: All the 3 datasets (e1,e2 & e3) are random, independent, parametric & normally distributed

Perform levene test on the data

The Levene test tests the null hypothesis that all input samples are from populations with equal variances. Levene’s test is an alternative to Bartlett’s test bartlett in the case where there are significant deviations from normality.

source: scipy.org

#### Answer:

In [75]:
lev = stats.levene(e1, e2, e3)
lev

LeveneResult(statistic=2.6741725711150446, pvalue=0.12259792666001798)

In [76]:
anov = stats.f_oneway(e1, e2, e3)
anov

F_onewayResult(statistic=2.51357622845924, pvalue=0.13574644501798466)

## Question 5

The one-way ANOVA tests the null hypothesis that two or more groups have the same population mean. The test is applied to samples from two or more groups, possibly with differing sizes.

use stats.f_oneway() module to perform one-way ANOVA test

In [77]:
anov = stats.f_oneway(e1, e2, e3)
anov

F_onewayResult(statistic=2.51357622845924, pvalue=0.13574644501798466)

## Question 6

*In one or two sentences explain about **TypeI** and **TypeII** errors.*

#### Answer:

Type 1 Error : Alpha ( $\alpha$ ): Reject the Null Hypothesis if its true. 

Type 2 Error : Beta ( $\beta$ ) : Failing to reject Null hypothesys if its False

## Question 7 

You are a manager of a chinese restaurant. You want to determine whether the waiting time to place an order has changed in the past month from its previous population mean value of 4.5 minutes. 
State the null and alternative hypothesis.

#### Answer:


* $H_0$: Null Hypothesis  : Waiting time to place an orderhas NOT changed in past 1 month. 

  * $\mu_0$ = 4.5

## Chi square test

## Question 8

Let's create a small dataset for dice rolls of four players

In [78]:
import numpy as np

d1 = [5, 8, 3, 8]
d2 = [9, 6, 8, 5]
d3 = [8, 12, 7, 2]
d4 = [4, 16, 7, 3]
d5 = [3, 9, 6, 5]
d6 = [7, 2, 5, 7]

dice = np.array([d1, d2, d3, d4, d5, d6])

run the test using SciPy Stats library

Depending on the test, we are generally looking for a threshold at either 0.05 or 0.01. Our test is significant (i.e. we reject the null hypothesis) if we get a p-value below our threshold.

For our purposes, we’ll use 0.01 as the threshold.

use stats.chi2_contingency() module 

This function computes the chi-square statistic and p-value for the hypothesis test of independence of the observed frequencies in the contingency table

Print the following:

- chi2 stat
- p-value
- degree of freedom
- contingency



In [79]:
chi2_stat, p_val, dof, ex = stats.chi2_contingency(dice)

In [80]:
print("\n Chi  = ", chi2_stat)

print("\n P - Value = ", p_val)

print("\n Degrees of Freedom = ", dof)

print("\n Contigency Table  \n", ex)


 Chi  =  23.315671914716496

 P - Value =  0.07766367301496693

 Degrees of Freedom =  15

 Contigency Table  
 [[ 5.57419355  8.20645161  5.57419355  4.64516129]
 [ 6.50322581  9.57419355  6.50322581  5.41935484]
 [ 6.73548387  9.91612903  6.73548387  5.61290323]
 [ 6.96774194 10.25806452  6.96774194  5.80645161]
 [ 5.34193548  7.86451613  5.34193548  4.4516129 ]
 [ 4.87741935  7.18064516  4.87741935  4.06451613]]


#### Since p-value == 7.77, we fail to reject the Null Hypothesis  

## Question 9

### Z-test

Get zscore on the above dice data using stats.zscore module from scipy. Convert zscore values to p-value and take mean of the array.

In [81]:
dice.mean()

6.458333333333333

In [82]:
stats.zscore(dice)

array([[-0.46291005, -0.18884739, -1.83711731,  1.44115338],
       [ 1.38873015, -0.64208114,  1.22474487,  0.        ],
       [ 0.9258201 ,  0.7176201 ,  0.61237244, -1.44115338],
       [-0.9258201 ,  1.62408759,  0.61237244, -0.96076892],
       [-1.38873015,  0.03776948,  0.        ,  0.        ],
       [ 0.46291005, -1.54854863, -0.61237244,  0.96076892]])

In [83]:
stats.norm.sf(stats.zscore(dice))

array([[0.67828558, 0.57489379, 0.96690371, 0.07477068],
       [0.08245741, 0.73958975, 0.11033568, 0.5       ],
       [0.17726974, 0.23649578, 0.27014569, 0.92522932],
       [0.82273026, 0.05217856, 0.27014569, 0.83166582],
       [0.91754259, 0.48493574, 0.5       , 0.5       ],
       [0.32171442, 0.93925487, 0.72985431, 0.16833418]])

In [84]:
stats.norm.sf(stats.zscore(dice)).mean()

0.49478056512575197

In [85]:
print("Mean p-Value : ", stats.norm.sf(stats.zscore(dice)).mean())

Mean p-Value :  0.49478056512575197


Since 0.495 > 0.01, we fail to reject the Null Hypothesis

In [86]:
# My Trial
stats.ttest_1samp(dice, 0)

Ttest_1sampResult(statistic=array([6.21059003, 4.47612146, 8.21583836, 5.37086156]), pvalue=array([0.00158119, 0.00654212, 0.00043498, 0.00301261]))

## Question 10

A Paired sample t-test compares means from the same group at different times.

The basic two sample t-test is designed for testing differences between independent groups. 
In some cases, you might be interested in testing differences between samples of the same group at different points in time. 
We can conduct a paired t-test using the scipy function stats.ttest_rel(). 

In [87]:
before= stats.norm.rvs(scale=30, loc=100, size=500) ## Creates a normal distribution with a mean value of 100 and std of 30
after = before + stats.norm.rvs(scale=5, loc=-1.25, size=500)

Test whether a weight-loss drug works by checking the weights of the same group patients before and after treatment using above data.

In [88]:
tt, pval = stats.ttest_rel(before, after)

print("T-Test   : ", tt)
print("p-Value  : ", pval)

T-Test   :  5.423946834682965
p-Value  :  9.102116914896419e-08


Since p-value is < 0.025, we reject the Null Hypothesis (that the waight loss is Not working). 